component/storage/mongodb/mongodb.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright Scoir Inc Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package mongodb implements a storage interface for Aries (aries-framework-go).
+//
+package mongodb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Data is a wrapper for the stored key/value pair.
+type Data struct {
+	Key   string `bson:"_id" json:"Key"`
+	Value []byte `bson:"Value" json:"Value"`
+}
+
+// Option configures the mongodb provider.
+type Option func(opts *Provider)
+
+// WithDBPrefix option is for adding prefix to db name.
+func WithDBPrefix(dbPrefix string) Option {
+	return func(opts *Provider) {
+		opts.dbPrefix = dbPrefix
+	}
+}
+
+// Provider mongodb implementation of storage.Provider interface.
+type Provider struct {
+	dial     string
+	dbs      map[string]*mongodbStore
+	dbPrefix string
+	sync.RWMutex
+}
+
+// NewProvider instantiates Provider.
+func NewProvider(dial string, opts ...Option) *Provider {
+	p := &Provider{dial: dial, dbs: map[string]*mongodbStore{}}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+// OpenStore opens and returns a store for given name space.
+func (r *Provider) OpenStore(name string) (storage.Store, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.dbPrefix != "" {
+		name = r.dbPrefix + "_" + name
+	}
+
+	store, ok := r.dbs[name]
+	if ok {
+		return store, nil
+	}
+
+	mongoClient, err := mongo.NewClient(options.Client().ApplyURI(r.dial))
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create new mongo client opening store")
+	}
+
+	err = mongoClient.Connect(context.Background())
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to connect to mongo opening a new store")
+	}
+
+	client := mongoClient
+	db := client.Database(name)
```

Comment:
Do we need extra reassignment to `client` ?

Reply:
Absolutely do not, good catch!  Fixed.